src/functions.rs

```diff
@@ -2740,6 +2750,22 @@ fn validate_argument<'a>(
             if arg_typeinfo.is_child_or_actual_type(target_argument.param_type, types) {
                 Ok(TypeInstance::new(&arg, arg_typeinfo, file, context))
             } else {
+                if let Some(file) = file {
+                    if file.name().is_empty() {
```

Comment:
Associated calls have an empty string file name?  That's not something I recall knowing, and probably not something that should be relied upon.

I believe FunctionInfo has an is_associated_call boolean field, which is the canonical way to know if something is an associated call.

I see that we don't have the FunctionInfo in validate_argument().  I suppose we *could* pass it in, but is validate_argument() even the right place to be checking?  Associated calls must have exactly one argument, so if we're validating the Nth argument for N > 1 then we don't need to do this check at all.  Maybe just pulling it up to the validate_arguments() level would make more sense?  Even then we don't have a FI.  Maybe even all the way up to ValidatedCall:new() actually.  We have an FI there, and the other callers of validate_arguments() are builtins, which can't be associated_calls so don't need this check.  That's not deeply thought through, so let me know if I'm missing something.

Reply:
Oh, I'm starting to track the logic better.  This is here right now, because we're in the case where validate_argument() has failed.  But since the associated_call was implicit, our generic error message is less than helpful.  So we need to put in a better error message.  The problem is that the FunctionInfo has all the information we need, and we don't have that here.

So the logic about moving it up to ValidatedCall as I outlined above would probably work, but seems redundant and a little weird.  This *is* the right place for error messages.  Maybe adding an Option<FunctionInfo> argument to validate_argument()/validate_arguments() would make sense, and then just passing None everywhere except ValidatedCall::new()?  I'm not wild about continuing to grow the argument list on those already bloated functions, but having the FunctionInfo we're calling seems like a reasonable thing to have while parsing arguments.